### PR TITLE
[Common][PyTorch] Add a new score func `sqrtsoftplus` to the fused router

### PR DESCRIPTION
**Pull request by yaox12**
# Description

- [x] Added a new score func `sqrtsoftplus`
- [x] Add tests
- [x] Switch to FP32 math instead of FP64 for better performance on SM103. Precision principles:
  - Inputs are casted into FP32 when loading from global memory.
  - All the math/calculations are in FP32 in the kernels.
  - `scores` is always in FP32 (match the MCore implementation).
  - `intermediate_output`  is always in FP32 for better backward precision.
  - Only cast to low-precision when necessary, for example, the gradient of input is required to have the same dtype as the input.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/csrc/extensions/router.cpp [hunk 1]
```diff
@@ -22,13 +23,14 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> fused_topk_with_score_function_fw
               "num_tokens and num_experts must be greater than 0");
   // Expert bias only happens at the sigmoid case
   if (expert_bias.has_value()) {
-    TORCH_CHECK(score_function == "sigmoid",
+    TORCH_CHECK(score_function == "sigmoid" || score_function == "sqrtsoftplus",
                 "score_function must be sigmoid when expert_bias is not None");
   }
```
[the remark this answers is not in the document]
Comment (yaox12) [reply transformer_engine/pytorch/csrc/extensions/router.cpp:30]: Updated.

File: transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu [hunk 2]
```diff
@@ -227,31 +233,53 @@ __global__ void fused_score_for_moe_aux_loss_backward_kernel(const DataType *int
...
+      double sum_Output_x_Grad = local_sum_Output_x_Grad;
       // In-place update
       for (int i = lane_id; i < num_experts; i += kThreadsPerWarp) {
         local_grad[i] =
             static_cast<double>(local_grad[i]) / (static_cast<double>(sum_fwd_input) + epsilon) -
-            static_cast<double>(sum_Output_x_Grad) /
-                ((static_cast<double>(sum_fwd_input) + epsilon) *
-                 (static_cast<double>(sum_fwd_input) + epsilon));
+            sum_Output_x_Grad / ((static_cast<double>(sum_fwd_input) + epsilon) *
+                                 (static_cast<double>(sum_fwd_input) + epsilon));
```
[the remark this answers is not in the document]
Comment (yaox12) [reply transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu]: Multiplied in Line 267.

File: transformer_engine/pytorch/csrc/extensions/router.cpp [hunk 3]
```diff
@@ -64,18 +68,14 @@ std::tuple<at::Tensor, at::Tensor, at::Tensor> fused_topk_with_score_function_fw
...
+void fused_topk_with_score_function_bwd(int num_tokens, int num_experts, at::Tensor routing_map,
+                                        at::Tensor intermediate_output, at::Tensor grad_probs,
+                                        at::Tensor grad_logits, int topk, bool use_pre_softmax,
+                                        std::optional<float> scaling_factor,
+                                        std::string score_function) {
   // Get the value of the parameters
   auto scaling_factor_value = scaling_factor.has_value() ? scaling_factor.value() : 1.0f;
   auto score_function_value = score_function_map[score_function];
-  // Init the output tensor
-  at::Tensor grad_logits = at::empty(
```
Comment (yaox12) [inline transformer_engine/pytorch/csrc/extensions/router.cpp:77]: Creating the output tensor on the Python side because the `intermediate_output` is now always in FP32, while the `grad_logits` should have the same dtype as the input.

File: transformer_engine/common/fused_router/fused_topk_with_score_function.cu [hunk 4]
```diff
@@ -15,16 +15,13 @@
 
 namespace transformer_engine {
 
-template <typename DataType, typename BiasType>
+template <typename DataType>
 __global__ void fused_topk_with_score_function_forward_kernel(
     const DataType *logits, int num_tokens, int num_experts, int topk, bool use_pre_softmax,
     int num_groups, int group_topk, float scaling_factor, int score_function,
-    const BiasType *expert_bias, DataType *probs, bool *routing_map,
```
Comment (denera) [inline transformer_engine/common/fused_router/fused_topk_with_score_function.cu:22]: Same as [here](https://github.com/NVIDIA/TransformerEngine/pull/2633/changes#r2854401192) for keeping `BiasType`.

File: transformer_engine/common/fused_router/fused_topk_with_score_function.cu [hunk 5]
```diff
@@ -33,24 +30,22 @@ __global__ void fused_topk_with_score_function_forward_kernel(
   int warp_id = threadIdx.x / kThreadsPerWarp;
   int lane_id = threadIdx.x % kThreadsPerWarp;
   extern __shared__ float shmem[];
-  DataType *scores_buf = reinterpret_cast<DataType *>(shmem);
-  DataType *topk_scores_buf =
-      reinterpret_cast<DataType *>(scores_buf + num_experts * num_token_per_block);
-  DataType *group_scores_buf = nullptr, *masked_scores_buf = nullptr;
+  float *scores_buf = shmem;
+  float *topk_scores_buf = scores_buf + num_experts * num_token_per_block;
+  float *group_scores_buf = nullptr, *masked_scores_buf = nullptr;
```
Comment (denera) [inline transformer_engine/common/fused_router/fused_topk_with_score_function.cu]: Same as [here](https://github.com/NVIDIA/TransformerEngine/pull/2633/changes#r2854381790) for `CompType` instead of `float`.

File: transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu [hunk 6]
```diff
@@ -33,13 +31,12 @@ __global__ void fused_score_for_moe_aux_loss_forward_kernel(const DataType *logi
   int warp_id = threadIdx.x / kThreadsPerWarp;
   int lane_id = threadIdx.x % kThreadsPerWarp;
   extern __shared__ float shmem_scores_for_aux_loss[];
-  DataType *logits_buf = reinterpret_cast<DataType *>(shmem_scores_for_aux_loss);
-  DataType *topk_logits_buf =
-      reinterpret_cast<DataType *>(logits_buf + num_experts * num_token_per_block);
+  float *logits_buf = shmem_scores_for_aux_loss;
+  float *topk_logits_buf = logits_buf + num_experts * num_token_per_block;
```
Comment (denera) [inline transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu]: We should alias these kinds of internal intermediate `float`s to `CompType` for consistency with other kernels.

File: transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu [hunk 7]
```diff
@@ -162,20 +164,19 @@ void fused_score_for_moe_aux_loss_forward(const Tensor &logits, int num_tokens,
...
           reinterpret_cast<bool *>(routing_map.data.dptr),
-          reinterpret_cast<DataType *>(intermediate_output.data.dptr), stream););
+          reinterpret_cast<float *>(intermediate_output.data.dptr), stream););
 }
 
 template <typename DataType>
-__global__ void fused_score_for_moe_aux_loss_backward_kernel(const DataType *intermediate_output,
-                                                             const DataType *grad_scores,
+__global__ void fused_score_for_moe_aux_loss_backward_kernel(const float *intermediate_output,
+                                                             const float *grad_scores,
```
Comment (denera) [inline transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu]: I don't think the kernel should make assumptions about the incoming data types here. We can do the internal compute in F32 but we should be able to work with whatever the kernel is invoked with instead of assuming every use case is going to keep their gradients in F32.

File: transformer_engine/common/fused_router/fused_topk_with_score_function.cu (hunk 4 above)
Comment (yaox12) [reply transformer_engine/common/fused_router/fused_topk_with_score_function.cu:22]: Changed accordingly.

File: transformer_engine/common/fused_router/fused_topk_with_score_function.cu (hunk 5 above)
Comment (yaox12) [reply transformer_engine/common/fused_router/fused_topk_with_score_function.cu]: Changed accordingly.

File: transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu (hunk 6 above)
Comment (yaox12) [reply transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu]: Changed accordingly.

File: transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu (hunk 7 above)
Comment (yaox12) [reply transformer_engine/common/fused_router/fused_score_for_moe_aux_loss.cu]: This function is a special case, where the output `scores` from forward function is always FP32, so  the gradient should also be in FP32.

**Comment by yaox12** [comment]
The failed `L0_cppunittest--B200_1GPU` is due to a CI infra error.

**Comment by KshitijLakhani** [comment]
@yaox12 FYI: https://github.com/NVIDIA/TransformerEngine/pull/2683#issuecomment-3976035763 
Thanks !
